xarray/core/nanops.py

```diff
@@ -0,0 +1,211 @@
+
+
+def _maybe_null_out(result, axis, mask, min_count=1):
+    """
+    xarray version of pandas.core.nanops._maybe_null_out
+    """
+    if hasattr(axis, '__len__'):  # if tuple or list
+        raise ValueError('min_count is not available for reduction '
+                         'with more than one dimensions.')
+
+    if axis is not None and getattr(result, 'ndim', False):
+        null_mask = (mask.shape[axis] - mask.sum(axis) - min_count) < 0
+        if np.any(null_mask):
+            dtype, fill_value = dtypes.maybe_promote(result.dtype)
+            result = result.astype(dtype)
+            result[null_mask] = fill_value
+
+    elif getattr(result, 'dtype', None) not in dtypes.NAT_TYPES:
+        null_mask = mask.size - mask.sum()
+        if null_mask < min_count:
+            result = np.nan
+
+    return result
+
+
+def _nan_argminmax_object(func, fill_value, value, axis=None, **kwargs):
+    """ In house nanargmin, nanargmax for object arrays. Always return integer
+    type """
+    valid_count = count(value, axis=axis)
+    value = fillna(value, fill_value)
+    data = getattr(np, func)(value, axis=axis, **kwargs)
+    # dask seems return non-integer type
+    if isinstance(value, dask_array_type):
+        data = data.astype(int)
+
+    if (valid_count == 0).any():
+        raise ValueError('All-NaN slice encountered')
+
+    return np.array(data, dtype=int)
+
+
+def _nan_minmax_object(func, fill_value, value, axis=None, **kwargs):
+    """ In house nanmin and nanmax for object array """
+    valid_count = count(value, axis=axis)
+    filled_value = fillna(value, fill_value)
+    data = getattr(np, func)(filled_value, axis=axis, **kwargs)
+    if not hasattr(data, 'dtype'):  # scalar case
+        data = dtypes.fill_value(value.dtype) if valid_count == 0 else data
+        return np.array(data, dtype=value.dtype)
+    return where_method(data, valid_count != 0)
+
+
+def nanmin(a, axis=None, out=None):
+    if a.dtype.kind == 'O':
+        return _nan_minmax_object(
+            'min', dtypes.get_pos_infinity(a.dtype), a, axis)
+
+    if isinstance(a, dask_array_type):
+        return dask_array.nanmin(a, axis=axis)
+    return nputils.nanmin(a, axis=axis)
+
+
+def nanmax(a, axis=None, out=None):
+    if a.dtype.kind == 'O':
+        return _nan_minmax_object(
+            'max', dtypes.get_neg_infinity(a.dtype), a, axis)
+
+    if isinstance(a, dask_array_type):
+        return dask_array.nanmax(a, axis=axis)
+    return nputils.nanmax(a, axis=axis)
+
+
+def nanargmin(a, axis=None):
+    fill_value = dtypes.get_pos_infinity(a.dtype)
+    if a.dtype.kind == 'O':
+        return _nan_argminmax_object('argmin', fill_value, a, axis=axis)
+    a, mask = _replace_nan(a, fill_value)
+    if isinstance(a, dask_array_type):
+        res = dask_array.argmin(a, axis=axis)
+    else:
+        res = np.argmin(a, axis=axis)
+
+    if mask is not None:
+        mask = np.all(mask, axis=axis)
+        if np.any(mask):
```

Comment:
It's more dask friendly to use the`.all()` and `.any()` methods -- that way at least most of the check can be done in parallel, and only a single value (the boolean scalar) gets evaluates.